lib/src/rules/parameter_assignments.dart

```diff
@@ -124,75 +106,95 @@ class ParameterAssignments extends LintRule {
+        rule.reportLint(node);
+      }
+    } else {
+      if (paramDefaultsToNull) {
+        if (_isFormalParameterReassigned(parameter, node)) {
+          if (hasBeenAssigned) {
+            rule.reportLint(node);
+          }
+          hasBeenAssigned = true;
         }
       }
     }
+
+    super.visitAssignmentExpression(node);
   }
 
   @override
-  void visitMethodDeclaration(MethodDeclaration node) {
-    var parameterList = node.parameters;
-    if (parameterList != null) {
-      // Getters don't have parameters.
-      for (var e in parameterList.parameters) {
-        var declaredElement = e.declaredElement;
-        if (declaredElement != null &&
-            node.body.isPotentiallyMutatedInScope(declaredElement)) {
-          _reportIfSimpleParameterOrWithDefaultValue(e, node);
-        }
+  visitPostfixExpression(PostfixExpression node) {
+    if (paramIsNotNullByDefault) {
+      var operand = node.operand;
+      if (operand is SimpleIdentifier &&
+          operand.staticElement == parameter.declaredElement) {
+        rule.reportLint(node);
       }
     }
+
+    super.visitPostfixExpression(node);
   }
 
-  void _reportIfSimpleParameterOrWithDefaultValue(
-      FormalParameter parameter, AstNode functionOrMethodDeclaration) {
-    var nodes = functionOrMethodDeclaration.traverseNodesInDFS();
-
-    if (parameter is SimpleFormalParameter ||
-        _isDefaultFormalParameterWithDefaultValue(parameter)) {
-      var mutatedNodes = nodes.where((n) =>
-          (n is AssignmentExpression &&
-              _isFormalParameterReassigned(parameter, n)) ||
-          _preOrPostFixExpressionMutation(parameter, n));
-      mutatedNodes.forEach(rule.reportLint);
-      return;
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    if (paramIsNotNullByDefault) {
+      var operand = node.operand;
+      if (operand is SimpleIdentifier &&
+          operand.staticElement == parameter.declaredElement) {
+        rule.reportLint(node);
+      }
     }
 
-    var assignmentsNodes = nodes
-        .where((n) =>
-            n is AssignmentExpression &&
-            _isDefaultFormalParameterWithoutDefaultValueReassigned(
-                parameter, n))
-        .toList();
-
-    var nonNullCoalescingAssignments = assignmentsNodes.where((n) =>
-        (n as AssignmentExpression).operator.type !=
-        TokenType.QUESTION_QUESTION_EQ);
-
-    if (assignmentsNodes.length > 1 ||
-        nonNullCoalescingAssignments.isNotEmpty) {
-      var node = assignmentsNodes.length > 1
-          ? assignmentsNodes.last
-          : nonNullCoalescingAssignments.isNotEmpty
-              ? nonNullCoalescingAssignments.first
-              : parameter;
-      rule.reportLint(node);
+    super.visitPrefixExpression(node);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitFunctionDeclaration(FunctionDeclaration node) {
```

Comment:
A `FunctionDeclaration` is a top-level function. We wrap it in a `FunctionDeclarationStatement` for local functions, so this method takes care of both of those cases. What it misses is closures, which are represented by `FunctionExpression`. If we should produce a diagnostic in the following case:
```
list.where((e) {
  e = 3; // Lint here?
});
```
then I think we need to visit `FunctionExpression` instead.

Reply:
Right.  Closures are not currently supported and adding them would be breaking so I'll add a TODO.